mlir/lib/Quantum/Transforms/annotate_function.cpp

```diff
@@ -0,0 +1,171 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "Quantum/IR/QuantumOps.h"
+#include "mlir/Analysis/CallGraph.h"
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "Catalyst/IR/CatalystOps.h"
+#include "Quantum/Transforms/Passes.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace {
+
+static constexpr const char *hasMeasureAttrName = "catalyst.invalidGradientOperation";
+
+bool isAnnotated(func::FuncOp op, const char *attr)
+{
+    return (bool)(op->getAttrOfType<UnitAttr>(attr));
+}
+
+bool invalidGradientOperation(func::FuncOp op)
+{
+    auto res = op.walk([](Operation *o) {
+        if (dyn_cast<MeasureOp>(o) || dyn_cast<catalyst::PythonCallOp>(o) ||
+            dyn_cast<catalyst::CustomCallOp>(o)) {
+            return WalkResult::interrupt();
+        }
+        else {
+            return WalkResult::advance();
+        }
+    });
+    return res.wasInterrupted();
+}
+
+bool successfulMatchLeaf(func::FuncOp op)
+{
+    return !isAnnotated(op, hasMeasureAttrName) && invalidGradientOperation(op);
+}
+
+void annotate(func::FuncOp op, PatternRewriter &rewriter, const char *attr)
+{
+    op->setAttr(attr, rewriter.getUnitAttr());
+}
+
+struct AnnotateFunctionTransform : public OpRewritePattern<func::FuncOp> {
+    using OpRewritePattern<func::FuncOp>::OpRewritePattern;
+
+    LogicalResult match(func::FuncOp op) const override;
+    void rewrite(func::FuncOp op, PatternRewriter &rewriter) const override;
+};
+
+LogicalResult AnnotateFunctionTransform::match(func::FuncOp op) const
+{
+    return successfulMatchLeaf(op) ? success() : failure();
+}
+
+void AnnotateFunctionTransform::rewrite(func::FuncOp op, PatternRewriter &rewriter) const
+{
+    annotate(op, rewriter, hasMeasureAttrName);
+}
+
+std::optional<func::FuncOp> getFuncOp(const CallGraphNode *node, CallGraph &cg)
+{
+    std::optional<func::FuncOp> funcOp = std::nullopt;
+    if (node == cg.getExternalCallerNode())
+        return funcOp;
+    if (node == cg.getUnknownCalleeNode())
+        return funcOp;
+    auto *callableRegion = node->getCallableRegion();
+    funcOp = cast<func::FuncOp>(callableRegion->getParentOp());
+    return funcOp;
+}
+
+std::optional<func::FuncOp> getCallee(CallGraphNode::Edge edge, CallGraph &cg)
+{
+    CallGraphNode *callee = edge.getTarget();
+    return getFuncOp(callee, cg);
+}
+
+bool anyCalleeIsAnnotated(func::FuncOp op, const char *attr, CallGraph &cg)
+{
+    Region &region = op.getRegion();
+    CallGraphNode *node = cg.lookupNode(&region);
+    // TODO: ICE if we do not find the node.
```

Comment:
What's ICE?

Reply:
Internal Compiler Error.

Reply:
Should we add that now?

Reply:
Mmm... I am always confused about which is the correct API to state this to the user and then end the compilation process. I don't think it is singalPassFailure but it might be.

Reply:
I think `signalPassFailure` indicates an unrecoverable state, which will most likely abort compilation entirely. If it is case that the pass simply doesn't apply and should ignore, then a match failure would be better. If it is a case that we don't expect though, and shouldn't exist, than the pass failure is appropriate I think :)

Reply:
We are already using `signalPassFailure` but it can only be used in the `Pass` class, not in the `Pattern` class, which is where the transformation happens. Essentially, what I would like to do is exit the pattern early and signal failure, but that is only possible through the `LogicalResult` which implies I catch the error during the `match` but not `rewrite` portion of the pattern. I think leaving it as is, is ok-ish. It would be better to emit a message and then abort. This is what I think `quantum-opt` should do, but not what the compiler driver should do. The compiler driver should likely raise an exception (?)

Reply:
In that case an exception is also a reasonable choice 👍

Reply:
Mmm... but this file is compiled with `-no-exceptions` :sweat_smile: 

```
/home/ubuntu/code/catalyst/mlir/lib/Quantum/Transforms/annotate_function.cpp:103:9: error: cannot use 'throw' with exceptions disabled
        throw std::runtime_error("Internal Compiler Error: This is a state that should not happen");
        ^
```

Reply:
Oh interesting. Assertions do work, but unfortunately they don't produce an error message in Python and instead just crash the process :/

One easy thing you can do to go with your above plan is to combine match and rewrite into the same step (`matchAndRewrite`), then you can also return a logical result from the "rewrite" step.
But I don't see how that would signal an error state, because "failure" just match the pattern didn't match which is not necessarily an error state 🤔

Reply:
I think it is fine to leave as is (or have an assertion an abort). This case would signal an error in MLIR's infrastructure.

Reply:
Ok, I just deleted the comment based on our discussion.

Reply:
Added the assertion https://github.com/PennyLaneAI/catalyst/pull/204/commits/0271f1e25e9d129abb0609d02843824d10bb5051